mockFn.go

```diff
@@ -0,0 +1,406 @@
+
+	mock.args = &args
+	return mock
+}
+
+// RecordCall is used by a mock implementation to record a call to a mock function,
+// optionally testing that arguments match those expected and returning the result and
+// error configured for the expected call.
+//
+// # returns
+//
+// Returns mock.expected.Result for an expected call; if there is no expected call
+// the zero value of the R type parameter is returned with ErrUnexpectedCall.
+//
+// If the arguments do not match the expected call, mock.expected.Result is
+// returned with ErrUnexpectedArgs.
+//
+// # errors
+//
+//	ErrUnexpectedCall       the call has no corresponding expected call
+//	ErrUnexpectedArgs       the arguments do not match the expected call
+//	mock.expected.Err	    the error specified for the expected call (if any)
+//
+// # example
+//
+//	type MyInterface interface {
+//		MyMethod(int) (int, error)
+//	}
+//
+//	type myMock struct {
+//		myMethod test.MockFn[int, int]
+//	}
+//
+//	func (mock *myMock) MyMethod(arg int) (int, error) {
+//		return mock.myMethod.RecordCall(arg)
+//	}
+//
+// # Multiple Return Values
+//
+// When a mocked function returns multiple values (in addition to an error), the
+// return value will typically be a struct with fields for each of the result values.
+//
+// The struct fields must be returned as individual return values by the mock
+// implementation:
+//
+//	func (mock *myMock) IntDiv(num, div int) (struct{Result, Remainder int}, error) {
+//		result, err := mock.intDiv.RecordCall(any)
+//		return result.Result, result.Remainder, err
+//	}
+func (mock *MockFn[A, R]) RecordCall(args ...A) (returns R, err error) {
+	if mock.responses != nil {
+		panic(fmt.Errorf("%w: mock function is configured for mapped results; use <fn>.ResultFor()", ErrInvalidOperation))
+	}
+
+	actual := &mockFnCall[A, R]{}
+	if len(args) > 0 {
+		actual.args = &args[0]
+	}
+
+	if mock.expected == nil {
+		actual.err = ErrUnexpectedCall
+		err = fmt.Errorf("%w: with args: %v", ErrUnexpectedCall, args)
+		mock.actual = append(mock.actual, actual)
+		mock.errs = append(mock.errs, err)
+		return
+	}
+
+	err = mock.expected.err
+	returns = mock.expected.result
+
+	switch {
+	case mock.expected.args == nil:
+		// NO-OP - no arguments of interest were configured for the expected call
+
+	case actual.args == nil:
+		actual.err = ErrExpectedArgs
+		err = fmt.Errorf("%w:\n  expected: %v\n  got     : nil (no args recorded)", ErrExpectedArgs, mock.expected.args)
+		mock.errs = append(mock.errs, err)
+
+	case *mock.expected.args != *actual.args:
+		actual.err = ErrUnexpectedArgs
+		err = fmt.Errorf("%w:\n  expected: %v\n  got     : %v", ErrUnexpectedArgs, mock.expected.args, actual.args)
+		mock.errs = append(mock.errs, err)
+	}
+
+	mock.actual = append(mock.actual, actual)
+	mock.idxExpected++
+	mock.expected = nil
+
+	if mock.idxExpected < len(mock.expectations) {
+		mock.expected = mock.expectations[mock.idxExpected]
+	} else {
+		mock.idxExpected = -1
+	}
+
+	return
+}
```

Comment:
Making mock function thread-safe would be inappropriate; mocks are not intended to be used in concurrent tests since without predictable serialisation of calls it would be impossible to configure expected calls correctly.

When running concurrent tests, the test author must simply take care to ensure that each test uses a separate and distinct mock.